rust/src/ssh/parser.rs

```diff
@@ -0,0 +1,187 @@
+/* Copyright (C) 2018 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+//nom5 use nom::bytes::complete::is_not;
+use nom::{be_u32, be_u8, rest};
+
+#[inline]
+fn is_not_lineend(b: u8) -> bool {
+    if b == 10 || b == 13 {
+        return false;
+    }
+    return true;
+}
+
+//may leave \r at the end to be removed
+named!(pub ssh_parse_line<&[u8], &[u8]>,
+    terminated!(
+        take_while!(is_not_lineend),
+        alt!( tag!("\n") | tag!("\r\n") |
+              do_parse!(
+                    bytes: tag!("\r") >>
+                    not!(eof!()) >> (bytes)
+                )
+            )
+    )
+);
+
+#[derive(PartialEq)]
+pub struct SshBanner<'a> {
+    pub protover: &'a [u8],
+    pub swver: &'a [u8],
+}
+
+impl<'a> SshBanner<'a> {}
+
+// Could be simplified adding dummy \n at the end
+// or use nom5 nom::bytes::complete::is_not
+named!(pub ssh_parse_banner<SshBanner>,
+    do_parse!(
+        tag!("SSH-") >>
+        protover: is_not!("-") >>
+        char!('-') >>
+        swver: alt!( complete!( is_not!(" \r") ) | rest ) >>
+        //remaning after space is comments
+        (SshBanner{protover, swver})
+    )
+);
+
+#[derive(PartialEq)]
+pub struct SshRecordHeader {
+    pub pkt_len: u32,
+    padding_len: u8,
+    pub msg_code: u8,
+}
+
+named!(pub ssh_parse_record_header<SshRecordHeader>,
+    do_parse!(
+        pkt_len: verify!(be_u32, |val:u32| val > 1) >>
+        padding_len: be_u8 >>
```

Comment:
Have tests been made against 0-length or 1-length packets without message code ?

Reply:
Not sure I understand what you're asking

Reply:
If I send a SSH record header with length == 0 or length == 1, the C and Rust parser will fail.
I am not sure how real implementations behave...
Should this be part of this PR ? Or should I create a new ticket for it (and keep this PR as only translating and not fixing bugs) ?

Reply:
Does the app-layer API ever send this up though?

Reply:
This is used by reaching the line `header->flags |= SSH_FLAG_PARSER_DONE;` which ends up doing `AppLayerParserStateSetFlag(pstate, APP_LAYER_PARSER_NO_INSPECTION);`

Reply:
Furthermore, we will need to get this right for https://redmine.openinfosecfoundation.org/issues/2698

Reply:
> This is used by reaching the line `header->flags |= SSH_FLAG_PARSER_DONE;` which ends up doing `AppLayerParserStateSetFlag(pstate, APP_LAYER_PARSER_NO_INSPECTION);`

So we're still calling the parser after this, but only with a 0 len payload?

Reply:
Oh, I think I understood four misunderstanding...

By 0-length or 1-length packets
I meant 0-length or 1-length _SSH records_

This ssh record begins by field `pkt_len` encoded on u32, then has a u8 padding and u8 msg_code
The `pkt_len` includes these u8.
Both C and Rust parsers forbid the use of `pkt_len < 2`
I do not know if real implementations allow and skip these dummy SSH records... That is my question.

Reply:
How hard would it to find out?

Reply:
A few hours I guess... Enough to ask if someone already did this ;-)